### PR TITLE
Another attempt to simplify SQL maintenance - option 2

### PR DESCRIPTION
**Pull request by agorina**
This is meant to be compared to https://github.com/SmarterApp/RDW_Reporting/pull/632.

### CODE REVIEW

File: common/src/main/java/org/opentestsystem/rdw/reporting/common/sqlbuilder/QueryProvider.java [hunk 1]
```diff
@@ -68,7 +69,7 @@ public String newQueryAsString(final String templateName, final Iterable<String>
 
     private void addClauses(final SqlTemplate template, final SelectQuery query) {
         // FROM clause is applicable to the main template only and is handled separately
-        final SqlClause[] clauses = {SELECT, JOIN, WHERE, GROUP_BY};
+        final SqlClause[] clauses = {SELECT, JOIN, WHERE, GROUP_BY, ORDER_BY};
```
Comment (agorina) [inline common/src/main/java/org/opentestsystem/rdw/reporting/common/sqlbuilder/QueryProvider.java]: This is unrelated, I was missing  `order by` support and an example query I choose needed it.

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcIabReportRepository.java [hunk 2]
```diff
@@ -131,11 +127,11 @@
...
             case Unknown:
-                return findAllIabsByStudentIdsAndQueryParamsUnknownPermissions;
+                return queryProvider.newQueryAsString("findAllExamsByStudentIdsAndQueryParams", ImmutableList.of("iab", "unknownScopePermission"));
             case Individual:
-                return findAllIabsByStudentIdsAndQueryParamsIndividualPermissions;
+                return queryProvider.newQueryAsString("findAllExamsByStudentIdsAndQueryParams", ImmutableList.of("iab", "individualScopePermission"));
             case Group:
-                return findAllIabsByStudentIdsAndQueryParamsGroupPermissions;
+                return queryProvider.newQueryAsString("findAllExamsByStudentIdsAndQueryParams", ImmutableList.of("iab", "groupScopePermission" ,"individualScopePermission"));
             default:
```
Comment (agorina) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcIabReportRepository.java]: This is how the final SQL is assembled from the `template`.

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcIabReportRepository.java [hunk 3]
```diff
@@ -131,11 +127,11 @@
...
         switch (queryType) {
             case Unknown:
-                return findAllIabsByStudentIdsAndQueryParamsUnknownPermissions;
+                return queryProvider.newQueryAsString("findAllExamsByStudentIdsAndQueryParams", ImmutableList.of("iab", "unknownScopePermission"));
             case Individual:
-                return findAllIabsByStudentIdsAndQueryParamsIndividualPermissions;
+                return queryProvider.newQueryAsString("findAllExamsByStudentIdsAndQueryParams", ImmutableList.of("iab", "individualScopePermission"));
             case Group:
-                return findAllIabsByStudentIdsAndQueryParamsGroupPermissions;
+                return queryProvider.newQueryAsString("findAllExamsByStudentIdsAndQueryParams", ImmutableList.of("iab", "groupScopePermission" ,"individualScopePermission"));
```
Comment (esotrope) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcIabReportRepository.java]: are the list of strings building the query?

Comment (agorina) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcIabReportRepository.java]: Yes, it `chains` the `addon's by the names on top of the main template SQL.

Comment (esotrope) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcIabReportRepository.java]: Wondering if we should build these on the fly or at repo construction time.
Also, have we come up with some criteria for when we dynamically build queries vs statically construct them with snippet substitution?
Seems like we would use the dynamic construction only when user input can cause a ton of permutations

Comment (agorina) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcIabReportRepository.java]: Yes, you are correct: we should use the dynamic construction only when the user input can cause a ton of permutations. This is not the case for this query. So I guess you are voting for option 1? Or just leave it as is? 

(I think they are all hard to read,not sure which way to go).

Comment (esotrope) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcIabReportRepository.java]: yeah for this i would go with snippets since the permutations are small, known and not directly influenced by user input. its good that we're testing out this cool new query toy though : j
